contracts/RewardPool.sol

```diff
@@ -256,6 +289,179 @@ contract RewardPool is AccessControl, ReentrancyGuard {
+
+      period.users.add(kpis[i].referrerAddress);
+      period.kpis[kpis[i].referrerAddress] = kpis[i].kpi;
+
+      emit KpiUpdated(
+        kpis[i].referrerAddress,
+        periodId,
+        kpis[i].kpi,
+        periodStart,
+        periodEndExclusive,
+        kpiFunctionId
+      );
+    }
+  }
+
+  /**
+   * @dev Allows the owner to process a reward period and distribute rewards
+   * @param periodStart Start of the reward period (unix timestamp)
+   * @param periodEndExclusive End of the reward period (unix timestamp)
+   * @param totalRewardAmount Total amount of rewards to distribute
+   * @notice Allowed only for address with DEFAULT_ADMIN_ROLE
+   */
+  function processPeriod(
+    uint48 periodStart,
+    uint48 periodEndExclusive,
+    uint256 totalRewardAmount
+  ) external onlyRole(DEFAULT_ADMIN_ROLE) {
+    bytes32 periodId = _getPeriodId(periodStart, periodEndExclusive);
+
+    PeriodData storage period = _periods[periodId];
+
+    if (period.processedAt != 0) revert PeriodAlreadyProcessed(periodId);
+
+    if (period.users.length() == 0)
+      revert PeriodInvalid(periodStart, periodEndExclusive);
+
+    IRewardFunction.Kpi[] memory kpis = new IRewardFunction.Kpi[](
+      period.users.length()
+    );
+
+    for (uint256 i = 0; i < period.users.length(); i++) {
+      address user = period.users.at(i);
+      uint256 kpi = period.kpis[user];
+
+      kpis[i] = IRewardFunction.Kpi({referrerAddress: user, kpi: kpi});
+    }
+
+    IRewardFunction.Reward[] memory rewards = IRewardFunction(
+      rewardFunctionAddress
+    ).calculateReward(kpis, totalRewardAmount);
+
+    uint256[] memory rewardFunctionArgs = new uint256[](2);
+    rewardFunctionArgs[0] = periodStart;
+    rewardFunctionArgs[1] = periodEndExclusive;
+
+    uint256 totalIssuedRewardAmount = 0;
+    uint256 numUsersRewarded = 0;
+
+    for (uint32 i = 0; i < rewards.length; i++) {
+      if (rewards[i].reward == 0) continue;
+
+      RewardData memory rewardData = RewardData({
+        user: rewards[i].referrerAddress,
+        amount: rewards[i].reward,
+        idempotencyKey: keccak256(
+          abi.encode(
+            rewards[i].referrerAddress,
+            periodStart,
+            periodEndExclusive
+          )
+        )
+      });
+
+      _addReward(rewardData, rewardFunctionArgs, i);
+      totalIssuedRewardAmount += rewards[i].reward;
+      numUsersRewarded++;
+    }
+
+    emit PeriodProcessed(
+      periodId,
+      periodStart,
+      periodEndExclusive,
+      totalRewardAmount,
+      totalIssuedRewardAmount,
+      numUsersRewarded
+    );
+
+    period.processedAt = uint48(block.timestamp);
+  }
+
+  /**
+   * @dev Returns the kpi of a user for a given period
+   * @param periodStart Start of the reward period (unix timestamp)
+   * @param periodEndExclusive End of the reward period (unix timestamp)
+   * @param user Address of the user
+   * @return uint256 kpi
+   */
+  function getPeriodKpi(
```

Comment:
Hmm, right now, if a user (referrer) doesn't have a KPI uploaded for a period, this will return 0; it will also return 0 if a referrer explicitly had a 0 uploaded as a KPI. It will _also_ return 0 if a period doesn't have _any_ data uploaded at all/doesn't exist. 

Should we differentiate between these cases? For example, should we check (using `getPeriodUsers`) if the given user exists in the specified period, and revert with a custom error if they're not present?